File: app/src/main/java/com/ruege/mobile/data/repository/PracticeRepository.java
package com.ruege.mobile.data.repository;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.ruege.mobile.data.local.dao.PracticeAttemptDao;
import com.ruege.mobile.data.local.dao.PracticeStatisticsDao;
import com.ruege.mobile.data.local.dao.TaskDao;
import com.ruege.mobile.data.local.entity.PracticeAttemptEntity;
import com.ruege.mobile.data.local.entity.PracticeStatisticsEntity;
import com.ruege.mobile.data.local.entity.TaskEntity;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import kotlinx.coroutines.flow.Flow;

public class PracticeRepository {
    private final PracticeAttemptDao practiceAttemptDao;
    private final PracticeStatisticsDao practiceStatisticsDao;
    private final TaskDao taskDao;
    private final Executor executor = Executors.newSingleThreadExecutor();

    public PracticeRepository(PracticeAttemptDao practiceAttemptDao, 
                             PracticeStatisticsDao practiceStatisticsDao,
                             TaskDao taskDao) {
        this.practiceAttemptDao = practiceAttemptDao;
        this.practiceStatisticsDao = practiceStatisticsDao;
        this.taskDao = taskDao;
    }

    // Методы для работы с попытками
    public Flow<List<PracticeAttemptEntity>> getAttemptsByTaskId(Integer taskId) {
        return practiceAttemptDao.getAttemptsByTaskId(taskId);
    }

    public Flow<List<PracticeAttemptEntity>> getAttemptsByEgeNumber(String egeNumber) {
        return practiceAttemptDao.getAttemptsByEgeNumber(egeNumber);
    }

    public Flow<List<PracticeAttemptEntity>> getRecentAttempts(int limit) {
        return practiceAttemptDao.getRecentAttempts(limit);
    }

    public Flow<Integer> getAttemptCountForTask(Integer taskId) {
        return practiceAttemptDao.getAttemptCountForTask(taskId);
    }

    public LiveData<Integer> getTotalAttempts() {
        // Используем MutableLiveData как временное решение
        MutableLiveData<Integer> result = new MutableLiveData<>(0);
        executor.execute(() -> {
            // В реальном приложении здесь должен быть код для получения данных из DAO
            result.postValue(0);
        });
        return result;
    }

    public LiveData<Integer> getTotalCorrectAttempts() {
        // Используем MutableLiveData как временное решение
        MutableLiveData<Integer> result = new MutableLiveData<>(0);
        executor.execute(() -> {
            // В реальном приложении здесь должен быть код для получения данных из DAO
            result.postValue(0);
        });
        return result;
    }

    // Методы для работы со статистикой
    public Flow<PracticeStatisticsEntity> getStatisticsByEgeNumber(String egeNumber) {
        return practiceStatisticsDao.getStatisticsByEgeNumber(egeNumber);
    }

    public Flow<List<PracticeStatisticsEntity>> getAllStatisticsByAttemptsDesc() {
        return practiceStatisticsDao.getAllStatisticsByAttemptsDesc();
    }

    public Flow<List<PracticeStatisticsEntity>> getAllStatisticsByRecentActivity() {
        return practiceStatisticsDao.getAllStatisticsByRecentActivity();
    }

    // Базовый метод для сохранения попыток
    public void saveAttempt(TaskEntity task, boolean isCorrect) {
        executor.execute(() -> {
            long currentTime = System.currentTimeMillis();
            
            // 1. Сначала сохраняем попытку
            PracticeAttemptEntity attempt = new PracticeAttemptEntity(
                    task.getId(), 
                    isCorrect, 
                    currentTime
            );
            practiceAttemptDao.insert(attempt);
            
            // 2. Убеждаемся, что запись статистики существует
            practiceStatisticsDao.createStatisticsIfNotExists(task.getEgeNumber());
            
            // 3. Обновляем статистику
            practiceStatisticsDao.updateStatisticsAfterAttempt(
                    task.getEgeNumber(),
                    isCorrect,
                    currentTime
            );
        });
    }
    
    // Расширенный метод для сохранения попыток с дополнительными параметрами
    public void saveAttempt(TaskEntity task, boolean isCorrect, String source, String taskType, String textId) {
        executor.execute(() -> {
            long currentTime = System.currentTimeMillis();
            
            // 1. Сначала сохраняем попытку
            PracticeAttemptEntity attempt = new PracticeAttemptEntity(
                    task.getId(), 
                    isCorrect, 
                    currentTime
            );
            practiceAttemptDao.insert(attempt);
            
            // 2. Убеждаемся, что запись статистики существует
            practiceStatisticsDao.createStatisticsIfNotExists(task.getEgeNumber());
            
            // 3. Обновляем статистику
            practiceStatisticsDao.updateStatisticsAfterAttempt(
                    task.getEgeNumber(),
                    isCorrect,
                    currentTime
            );
            
            // Логирование дополнительных параметров можно добавить при необходимости
        });
    }

    // Получение статистики выполнения по всем заданиям
    public LiveData<List<PracticeStatisticsEntity>> getStatisticsWithAttempts() {
        // Используем MutableLiveData как временное решение
        MutableLiveData<List<PracticeStatisticsEntity>> result = new MutableLiveData<>(new ArrayList<>());
        executor.execute(() -> {
            // В реальном приложении здесь должен быть код для получения данных из DAO
            result.postValue(new ArrayList<>());
        });
        return result;
    }
    
    /**
     * Получает задание по его идентификатору
     * @param taskId идентификатор задания
     * @return задание или null, если не найдено
     */
    public TaskEntity getTaskById(Integer taskId) {
        return taskDao.getTaskByIdSync(taskId);
    }
} 

File: app/src/main/java/com/ruege/mobile/data/repository/PracticeRepository.kt
package com.ruege.mobile.data.repository

import androidx.lifecycle.LiveData
import androidx.lifecycle.liveData
import androidx.lifecycle.asLiveData
import com.ruege.mobile.data.local.dao.PracticeAttemptDao
import com.ruege.mobile.data.local.dao.PracticeStatisticsDao
import com.ruege.mobile.data.local.dao.TaskDao
import com.ruege.mobile.data.local.entity.PracticeAttemptEntity
import com.ruege.mobile.data.local.entity.PracticeStatisticsEntity
import com.ruege.mobile.data.local.entity.TaskEntity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext

class PracticeRepository(
    private val practiceAttemptDao: PracticeAttemptDao,
    private val practiceStatisticsDao: PracticeStatisticsDao,
    private val taskDao: TaskDao
) {

    fun getRecentAttempts(limit: Int): Flow<List<PracticeAttemptEntity>> {
        return practiceAttemptDao.getRecentAttempts(limit)
    }

    fun getTotalAttempts(): LiveData<Int> = liveData(Dispatchers.IO) {
        practiceStatisticsDao.getOverallAggregatedStatistics().map { aggStats ->
            aggStats?.totalAttempts ?: 0
        }.collect { emit(it) }
    }

    fun getTotalCorrectAttempts(): LiveData<Int> = liveData(Dispatchers.IO) {
        practiceStatisticsDao.getOverallAggregatedStatistics().map { aggStats ->
            aggStats?.correctAttempts ?: 0
        }.collect { emit(it) }
    }

    suspend fun saveAttempt(task: TaskEntity, isCorrect: Boolean, source: String, taskType: String, textId: String) {
        withContext(Dispatchers.IO) {
            val currentTime = System.currentTimeMillis()

            val attempt = PracticeAttemptEntity(
                task.id,
                isCorrect,
                currentTime
            )
            practiceAttemptDao.insert(attempt)
            
            task.egeNumber?.let { egeNum ->
                practiceStatisticsDao.createStatisticsIfNotExists(egeNum)
                practiceStatisticsDao.updateStatisticsAfterAttempt(
                    egeNum,
                    isCorrect,
                    currentTime
                )
            }
        }
    }

    fun getStatisticsWithAttempts(): LiveData<List<PracticeStatisticsEntity>> {
        return practiceStatisticsDao.getStatisticsWithAttempts().asLiveData()
    }

    fun getEssayStatistics(): LiveData<List<PracticeStatisticsEntity>> {
        return practiceStatisticsDao.getEssayStatistics().asLiveData()
    }

    fun getVariantStatistics(): LiveData<List<PracticeStatisticsEntity>> {
        return practiceStatisticsDao.getVariantStatistics().asLiveData()
    }

    /**
     * Получает задание по его идентификатору
     * @param taskId идентификатор задания
     * @return задание или null, если не найдено
     */
    suspend fun getTaskById(taskId: Int): TaskEntity? {
        return withContext(Dispatchers.IO) {
            taskDao.getTaskByIdSync(taskId)
        }
    }
} 